### ISSUE TITLE
How accurate is "Average processing time"?

### ISSUE THREAD

**Issue by jslawler** [AdguardTeam/AdGuardHome#691]
<img width="590" alt="Screen Shot 2019-03-30 at 10 28 27" src="https://user-images.githubusercontent.com/6005161/55268566-c40e2c00-5282-11e9-8536-b3cda76d7c9a.png">

Under Dashboard > General statistics my average processing time ranges between 380ms and 4000ms.

- What factors come into play with this calculation?
- Is this what you'd call normal statistics?
- What information would you need from me to estimate the normality of these statistics?

**Comment by jslawler** [comment]
I'd say for the most part this can be ignored, when analysing the query log there were a lot of outliers. As a discussion point, would it be worth having an advanced filtering section do you think where we could filter out latency outliers as well as many other things?

**Comment by ameshkov** [comment]
> What factors come into play with this calculation?

This is the average time spent on processing DNS queries starting at the moment when it enters AGH.

> As a discussion point, would it be worth having an advanced filtering section do you think where we could filter out latency outliers as well as many other things?

Well, for one, we should add the "Elapsed time" column to the query log, or maybe even add detailed timing there (i.e. see timing in Chrome's developer tools).

What else would you suggest?

**Comment by saltama** [comment]
A good measurement for the "normality" is the mode i suppose (instead of the simple average that doesn't consider how many data points are near that value). In addition, it could make sense to add the standard deviation or something similar. But hey, not sure if the average user will be able to interpret that.

The first time i saw that my reaction was that it was impossible to understand how much of that time is due to upstream dns resolution and how much to the filtering adguard does (my guess is that the filtering is quite fast and just a small part of the total).

**Comment by jslawler** [comment]
> This is the average time spent on processing DNS queries starting at the moment when it enters AGH.

Okay, so quite a simple calculation with no fancy additions. Thanks.

> Well, for one, we should add the "Elapsed time" column to the query log, or maybe even add detailed timing there (i.e. see timing in Chrome's developer tools).

I like the way you're heading there however I feel we're moving into a field now where AGH is going to need a basic and detailed view in the Query Log (and eventually other sections).

If you do add this data view, I can see a lot of other requests being made for viewable data so it might end up being easier in the long run to create a system within AGH for users to create their own data exports based off wildcards.

I'll raise these 2 suggestions as issues regardless so they're not lost.

> What else would you suggest?

I think a [trimmed mean](https://www.easycalculation.com/statistics/learn-trimmed-mean.php) should come into play for the average processing time, instead of simply the mean of all DNS requests.

Basically, trim the top and bottom (for theories sake)20% of DNS request time data and give the mean on the remainder. Extra thought will need to be put into this to factor in all the possible outcomes though.

I'll raise the idea as an issue for discussion as I think this is something that could be improved on.

**Comment by ameshkov** [comment]
Let's continue the discussion here and then file feature requests / tasks.

Just throwing ideas for now, here's what can be possibly useful:

1. Min, max, stddev, median values (not sure how to display all these without users freaking out).
2. Stats per upstream
3. Detailed timing (what part of the time was spent on filtering, querying upstream, filtering, querying browsing security and parental control web services)

Also, there was a feature request about prometheus endpoint, it seems relevant to this discussion: #516 .

**Comment by jslawler** [comment]
> Let's continue the discussion here and then file feature requests / tasks.
> 
> Just throwing ideas for now, here's what can be possibly useful:
> 
>     Min, max, stddev, median values (not sure how to display all these without users freaking out).
>     Stats per upstream
>     Detailed timing (what part of the time was spent on filtering, querying upstream, filtering, querying browsing security and parental control web services)
> 
> Also, there was a feature request about prometheus endpoint, it seems relevant to this discussion: #516 .

In regards to calculating processing time I think simplification should be the focus when designing a product to cater to the the general public.

With simplicity in mind, having a 1-line number like you do now is about as simple as it gets, however we are seeing inaccuracies.
After thinking on this a bit I can think of 2 ways to attack this and they can both actually work together.

1. A 1-line calculation based off all the values you've mentioned with a hover overlay (?) of the statistic breakdown for that calculation.
2. A line graph displaying x=query processing time(ms), y=interval (selectable by user). Of course, a calculation would have to come into play here as each interval on the graph could potentially have thousands of queries. The 1-line calculation would come in handy here as well, not only displaying the details from the 1-line calculation but the amount of queries at that time as well.

<img width="569" alt="Screen Shot 2019-04-02 at 19 47 02" src="https://user-images.githubusercontent.com/6005161/55393267-79afd800-552c-11e9-86a3-b0358e53dbb9.png">

I personally like the combination of both these ideas together, it's user friendly and provides all the extra information for the tech nerds.

Thoughts?

**Comment by ameshkov** [comment]
Generally, I like the idea. Doing all this will require a serious rework of the stats and the dashboard so we should do it in smaller steps.

Here's what I see in the end (this is the ideal scenario, I just want to see the end goal at this point):

1. Calculate detailed stats per upstream (requests count, detailed timing, avg/min/max/stdev values)
2. `DNS Queries` value should have a hover-triggered popover with per-upstream queries count. For example:
    ```
    tls://1.1.1.1: 123123
    tls://1.0.0.1: 32121
    ```
3. Avg. processing time should also have a popover with smth like this (I am listing everything here, but we'll need to find a way to fit it properly):
    ```
    tls://1.1.1.1 latency: 123ms
    tls://1.0.0.1 latency: 321ms
    Browsing security web service: 123ms
    Parental control web service: 422ms
    Filtering DNS queries: 1ms
    ```

4. Add "processing time" chart to the dashboard.
   
    We could also unite the current four small charts in a single one, smth like this: https://uploads.adguard.com/up04_7pvnw_Area_Chart_Example_-_JSFiddle.png

**Comment by jslawler** [comment]
> Generally, I like the idea. Doing all this will require a serious rework of the stats and the dashboard so we should do it in smaller steps.
> 
> Here's what I see in the end (this is the ideal scenario, I just want to see the end goal at this point):
> 
> 1. Calculate detailed stats per upstream (requests count, detailed timing, avg/min/max/stdev values)
> 2. `DNS Queries` value should have a hover-triggered popover with per-upstream queries count. For example:
>    ```
>    tls://1.1.1.1: 123123
>    tls://1.0.0.1: 32121
>    ```
> 3. Avg. processing time should also have a popover with smth like this (I am listing everything here, but we'll need to find a way to fit it properly):
>    ```
>    tls://1.1.1.1 latency: 123ms
>    tls://1.0.0.1 latency: 321ms
>    Browsing security web service: 123ms
>    Parental control web service: 422ms
>    Filtering DNS queries: 1ms
>    ```
> 4. Add "processing time" chart to the dashboard.
>    We could also unite the current four small charts in a single one, smth like this: https://uploads.adguard.com/up04_7pvnw_Area_Chart_Example_-_JSFiddle.png

I agree on all points, and I know the UI designer is going to do a fantastic Hulk green job with it. ;)

It was never going to be an easy solution, but rather a long term solution I could definitely see AGH ending up with in the end (depending on what the end result desired is, AIO package?).

I'm happy with that result - unless anyone else wants to chime in, I'd consider this ready to be give a label/priority.

**Comment by ameshkov** [comment]
Well, the only problem I see with this is fitting other modules statistics (we're planning to add mitm proxy, and possibly a VPN/shadowsocks for those who're running AGH on VPS). Although, We will need to straighten up the DNS stats anyway.

If there will be no new comments or ideas about this in the next couple of days, I'll split this issue into a number of smaller tasks.

**Comment by LordMerlin** [comment]
I'll keep the conversation going.
What kind of strange behavior? Why such a huge time lag?
"Average processing time | 129215.53"

**Comment by jslawler** [comment]
> I'll keep the conversation going.
> What kind of strange behavior? Why such a huge time lag?
> "Average processing time | 129215.53"

There are a lot of factors that come into play with this. As it's an average processing time a high latency is not a direct correlation with an issue.
For example, let's say you have 5 requests:

- 20ms
- 23ms
- 10ms
- 50ms
- 20,000ms
The average of these 5 requests would be 4103ms. Here you see the average processing time appears high, however likely due to 1 timeout request or slow request, when removed there's actually no issue at all with the performance of, in this case, AdGuard Home.

**Comment by ameshkov** [comment]
@jslawler well, yeah, but the default timeout is 10 sec, so if it shows 130sec avg, there is a case when the request hasn't timed out properly for some reason.

@LordMerlin if you run AGH with `-v` argument, you'll see lines with `RTT: xxxx` in the log. Please look for unusually high values there.
